compiler/rustc_hir_analysis/src/check/compare_method.rs

```diff
@@ -555,19 +555,25 @@ pub fn collect_trait_impl_trait_tys<'tcx>(
                     // contains `def_id`'s early-bound regions.
                     let id_substs = InternalSubsts::identity_for_item(tcx, def_id);
                     debug!(?id_substs, ?substs);
-                    let map: FxHashMap<ty::GenericArg<'tcx>, ty::GenericArg<'tcx>> = substs
-                        .iter()
-                        .enumerate()
-                        .map(|(index, arg)| (arg, id_substs[index]))
-                        .collect();
+                    let map: FxHashMap<ty::GenericArg<'tcx>, ty::GenericArg<'tcx>> =
+                        std::iter::zip(substs, id_substs).collect();
                     debug!(?map);
 
+                    // HACK(RPITIT): We want this type's regions to act like early bound regions
+                    // on the impl's method generics, not the trait's method generics, so adjust
+                    // the index of each early bound region to account for the fact that the trait
+                    // and impl method substs have different lengths.
                     let ty = tcx.fold_regions(ty, |region, _| {
-                        if let ty::ReFree(_) = region.kind() {
-                            map[&region.into()].expect_region()
-                        } else {
-                            region
-                        }
+                        let ty::ReFree(_) = region.kind() else { return region; };
+                        let ty::ReEarlyBound(e) = map[&region.into()].expect_region().kind()
+                            else { bug!("expected ReFree to map to ReEarlyBound"); };
+                        tcx.mk_region(ty::ReEarlyBound(ty::EarlyBoundRegion {
+                            def_id: e.def_id,
+                            name: e.name,
+                            index: (e.index as usize - trait_to_impl_substs.len()
+                                + tcx.generics_of(impl_m.container_id(tcx)).params.len())
+                                as u32,
+                        }))
                     });
                     debug!(%ty);
```

Comment:
While trying to understand what happens, I remarked that this scheme leaks inference variables.
For instance, in `src/test/ui/async-await/in-trait/issue-102310.rs`, you get `ty=Opaque(DefId(0:14 ~ issue_102310[ed7a]::{impl#0}::transaction::{opaque#0}), [F, R, '_#0r])`.

Reply:
Fixed by equating the full signature instead of just equating the return types, which allows us to infer things fully.

Reply:
Also #102903 asserts against this more strongly, since `fully_resolve` really shouldn't be returning inference variables..